### PR TITLE
settings: made monitors selectable in fullscreen modes

### LINKED ISSUE #12363: Allow choosing which monitor the game should use, on multi-monitor systems
This requires a bit of extra SDL plumbing to do properly, but is straightforward.

### PR DESCRIPTION
**Pull request by everclear72216**
An integer value has been added to the graphics settings and is
pushed though the Platform interface to the Sdl2PlatformWindow.
The integer represents the display index used by SDL to identify
a display device.

The platform window hinterface has been extended to provide a
method which can be used to retrieve the available monitors.
The renderer class has been extended to give access to this
information within the settings logic class.

In the display tab of the settings menu the monitor can be
selected via a dropdown iff a non-windowed mode is selected.

"restart necessary"-logic has been updated to cater for the new
configuration option which can only be applied after restart.

The UI has been adaptet for mods/common and mods/cnc

closes #12363

Thank you for your contribution to OpenRA!

Please be aware that we do not have enough project maintainers to match the rate of contributions, so it may take several days before somebody is able to respond to your Pull Request.

You can help speed up the review process by following a few steps:

* Make sure that you have read and understand the OpenRA Coding Standard (see https://github.com/OpenRA/OpenRA/wiki/Coding-Standard).
* Write quality commit messages (see https://chris.beams.io/posts/git-commit/).
* Only commit changes that directly relate to your Pull Request.  Use your Git interface to unstage any unrelated changes to project files, line endings, whitespace, or other files.
* Review the code diff view below to double check that your changes satisfy the above three points.
* Use the `make test` and `make check` commands to check for (and fix!) any issues that are reported by our automated tests.
* If you are changing shared mod or engine code, make sure that you have tested your changes in all four default mods.
* Respond to review comments as soon as you reasonably can.  Reviewers will usually prioritize Pull Requests that are still fresh in their minds.  Make sure to leave a comment when you push new changes, otherwise GitHub does not automatically notify reviewers!
* Leave a polite comment asking for reviews if a week or more has passed without feedback.

If you need any help you can ask in the #openra IRC channel on freenode (most active during European evenings).

### CODE REVIEW

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 1]
```diff
@@ -450,5 +453,22 @@ public void SetScaleModifier(float scale)
 			scaleModifier = scale;
 			OnWindowScaleChanged(windowScale, windowScale * oldScaleModifier, windowScale, windowScale * scaleModifier);
 		}
+
+		public IReadOnlyDictionary<string, int> GetDisplays()
```
Comment (teinarss) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Imo better to have this method returning a list of `Pair<string, int>`.

**Review by teinarss — DISMISSED** [review]

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs (hunk 1 above)
Comment (pchote) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Is there a reason why this cant return a string[]?

Comment (teinarss) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Its using the number here later in the call stack https://github.com/OpenRA/OpenRA/pull/17648/files#diff-e2bdca10e38c0db34ae3c2d91cd2fc48R136

Comment (pchote) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Yeah, but unless im missing something the number is just the array index?

Comment (teinarss) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: true

Comment (everclear72216) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: I agree, the change to an array is no problem.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 2]
```diff
@@ -830,6 +839,23 @@ static void ShowStatusBarsDropdown(DropDownButtonWidget dropdown, GameSettings s
...
+		{
+			var options = Game.Renderer.Displays;
+
+			Func<string, ScrollItemWidget, ScrollItemWidget> setupItem = (o, itemTemplate) =>
+			{
+				var item = ScrollItemWidget.Setup(itemTemplate,
+					() => s.VideoDisplay == options[o],
+					() => s.VideoDisplay = options[o]);
+
+				item.Get<LabelWidget>("LABEL").GetText = () => o;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Please use something like
```csharp
var font = Game.Renderer.Fonts[itemTemplate.Font];
var label = WidgetUtils.TruncateText(o, itemTemplate.Bounds.Width, font);
item.Get<LabelWidget>("LABEL").GetText = () => label
```
to protect against long display names.

We'll need to do a similar cut (using `CachedTransform` - there are several examples elsewhere in the UI code for this) for the selected dropdown display.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 3]
```diff
@@ -830,6 +839,23 @@ static void ShowStatusBarsDropdown(DropDownButtonWidget dropdown, GameSettings s
...
+				return item;
+			};
+
+			dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 500, options.Keys, setupItem);
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: The `options.Keys` here can change to `Enumerable.Range(0, options.Length)` to compensate for changing the type of `Game.Renderer.Displays`.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: I would define this as
```csharp
public string[] GetDisplayNames()
{
	return Exts.MakeArray(SDL.SDL_GetNumVideoDisplays(), SDL.SDL_GetDisplayName);
}
```
(and change the interface/Renderer method names to match).

The key changes being:
* Changing the type as mentioned above
* Returning the unmodified names (the number is a UI display detail that should be handled by the SettingsLogic)
* Much simpler code

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 4]
```diff
@@ -92,11 +93,13 @@ public Size SurfaceSize
 		[DllImport("user32.dll")]
 		static extern bool SetProcessDPIAware();
 
-		public Sdl2PlatformWindow(Size requestEffectiveWindowSize, WindowMode windowMode, float scaleModifier, int batchSize)
+		public Sdl2PlatformWindow(Size requestEffectiveWindowSize, WindowMode windowMode, float scaleModifier, int batchSize, int videoDisplay)
 		{
 			// Lock the Window/Surface properties until initialization is complete
 			lock (syncObject)
 			{
+				SDL.SDL_VideoInit(SDL.SDL_GetCurrentVideoDriver());
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Why is this needed? The `SDL_Init` line just below should already be doing this.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 5]
```diff
@@ -130,7 +133,7 @@ public Sdl2PlatformWindow(Size requestEffectiveWindowSize, WindowMode windowMode
 				Console.WriteLine("Using SDL 2 with OpenGL{0} renderer", useGLES ? " ES" : "");
 
 				SDL.SDL_DisplayMode display;
-				SDL.SDL_GetCurrentDisplayMode(0, out display);
+				SDL.SDL_GetCurrentDisplayMode(videoDisplay, out display);
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:152]: The `SDL_GetDisplayDPI` call just below will also need to use `videoDisplay`, and the workaround for macOS on lines 243/244 will need to be removed (the problem it was working around doesn't seem to occur anymore, and it otherwise forces the game back to the left-most monitor).

**Review by pchote — COMMENTED** [review]
Looks pretty good already, just a few smallish comments:

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs (hunk 1 above)
Comment (teinarss) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: My concern with using `string[]` is that now we are using the arrays index to keep track on which index the monitor is on. It will make the code a little more brittle. For example if someone would change the order of how the items in the dropdown a presented by changing the order of the list it would introduce a bug. Anyway don't think it will be any problems here tho so nothing that stop this from proceed.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs (hunk 2 above)
Comment (teinarss) [reply OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Probably better to use "Display N" here or something. Cant tell which display is which :)
display

Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: macOS gives the actual display model (e.g U2718Q), but if this is what we have to deal with under windows then I completely agree about changing these to "Display N". Its not worth the effort to present "Generic PnP M... (1)".

Comment (everclear72216) [reply OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Linux also gives back proper names for the displays. So which version would you like? Naming used here in the comments also raises the question whether it should be called "display" or "monitor".

Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: ~70% of players are on Windows, so lets skip the SDL-provided names. MacOS and I think Windows (someone correct me if i'm wrong) uses Display, so I think we should go with that for consistency.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs (hunk 5 above)
Comment (everclear72216) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:152]: I removed the code that moves the window back to the origin on MACs.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs (hunk 4 above)
Comment (everclear72216) [reply OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: It is not, thx.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 6]
```diff
@@ -250,13 +259,18 @@ Action InitDisplayPanel(Widget panel)
...
+				ds.VideoDisplay != OriginalVideoDisplay);
 
 			var statusBarsDropDown = panel.Get<DropDownButtonWidget>("STATUS_BAR_DROPDOWN");
 			statusBarsDropDown.OnMouseDown = _ => ShowStatusBarsDropdown(statusBarsDropDown, gs);
 			statusBarsDropDown.GetText = () => gs.StatusBars == StatusBarsType.Standard ?
 				"Standard" : gs.StatusBars == StatusBarsType.DamageShow ? "Show On Damage" : "Always Show";
 
+			var displaySelectionDropDown = panel.Get<DropDownButtonWidget>("DISPLAY_SELECTION_DROPDOWN");
+			displaySelectionDropDown.OnMouseDown = _ => ShowDisplaySelectionDropdown(displaySelectionDropDown, ds);
+			displaySelectionDropDown.GetText = () => string.Format("{0} {1}", DisplayBaseName, ds.VideoDisplay);
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Style nits: OpenRA code style is to use `"Display {0}".F(ds.VideoDisplay)`, and this should also be wrapped in a `CachedTransform<int, string>` keyed on `ds.VideoDisplay` to avoid recreating this string 60 times per second.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 7]
```diff
@@ -289,6 +303,7 @@ Action InitDisplayPanel(Widget panel)
 
 			uiScaleDropdown.IsDisabled = () => disableUIScale;
 
+			panel.Get("SCREEN_SELECTION").IsVisible = () => ds.Mode != WindowMode.Windowed;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Style nit: Please rename this to `DISPLAY_SELECTION` to match `DISPLAY_SELECTION_DROPDOWN`

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 8]
```diff
@@ -50,8 +53,13 @@ enum PanelType { Display, Audio, Input, Hotkeys, Advanced }
 		static SettingsLogic()
 		{
 			var original = Game.Settings;
+
+			// hardware changes are able to invalidate the stored value so get the current display
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Style nit: Please start comments with capital letters (also applies to other comments in the PR).

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 9]
```diff
@@ -830,6 +848,23 @@ static void ShowStatusBarsDropdown(DropDownButtonWidget dropdown, GameSettings s
...
 		}
 
+		static void ShowDisplaySelectionDropdown(DropDownButtonWidget dropdown, GraphicSettings s)
+		{
+			var options = Exts.MakeArray(Game.Renderer.DisplayCount, i => string.Format("{0} {1}", DisplayBaseName, i));
+
+			Func<string, ScrollItemWidget, ScrollItemWidget> setupItem = (o, itemTemplate) =>
+			{
+				var item = ScrollItemWidget.Setup(itemTemplate,
+					() => s.VideoDisplay == options.IndexOf(o),
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Using `IndexOf` to work backwards to find the index isn't great - please use my previous suggestion to pass the integer indices to `setupItem` instead.

IMO the label text should also be using the index + 1 instead of counting from 0.

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs [hunk 10]
```diff
@@ -129,16 +145,19 @@ public Sdl2PlatformWindow(Size requestEffectiveWindowSize, WindowMode windowMode
 
 				Console.WriteLine("Using SDL 2 with OpenGL{0} renderer", useGLES ? " ES" : "");
 
+				if (videoDisplay >= DisplayCount)
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: We should also check for negative values here.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs (hunk 2 above)
Comment (everclear72216) [reply OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: > Please use something like
> 
> ```cs
> var font = Game.Renderer.Fonts[itemTemplate.Font];
> var label = WidgetUtils.TruncateText(o, itemTemplate.Bounds.Width, font);
> item.Get<LabelWidget>("LABEL").GetText = () => label
> ```
> 
> to protect against long display names.
> 
> We'll need to do a similar cut (using `CachedTransform` - there are several examples elsewhere in the UI code for this) for the selected dropdown display.

I think this is resolved due to the simpler approach with a constant string which cannot result in options that are too long.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs (hunk 9 above)
Comment (everclear72216) [reply OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Changed it to count from 1.

Comment (everclear72216) [reply OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: I admit I did not understand what you were getting at. Now I do and it is done.

**Comment by pchote** [comment]
It looks like you maybe forgot to push the latest fixes?

**Comment by everclear72216** [comment]
Sorry, I was just hesitant because i forgot to test tiberian dawn, which has a different settings.yaml.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 11]
```diff
@@ -50,8 +53,13 @@ enum PanelType { Display, Audio, Input, Hotkeys, Advanced }
 		static SettingsLogic()
 		{
 			var original = Game.Settings;
+
+			// Saved settings may have been invalidated by a hardware change
+			original.Graphics.VideoDisplay = Game.Renderer.CurrentDisplay;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: https://github.com/OpenRA/OpenRA/blob/de0bb9ee39b96c8db46e3a76419d8fc519d3c99c/OpenRA.Mods.Common/LoadScreens/BlankLoadScreen.cs#L113-L122 would be a better place to do this IMO - otherwise we only update the setting when the player opens the settings menu.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 12]
```diff
@@ -373,6 +389,9 @@ Action ResetDisplayPanel(Widget panel)
 				ds.CursorDouble = dds.CursorDouble;
 				ds.ViewportDistance = dds.ViewportDistance;
 
+				// Saved settings may have been invalidated by a hardware change
+				ds.VideoDisplay = Game.Renderer.CurrentDisplay;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: The expected behaviour for the reset button is to move the window to the primary display after a restart - so this should be `ds.VideoDisplay = dds.VideoDisplay`.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 13]
```diff
@@ -830,6 +849,23 @@ static void ShowStatusBarsDropdown(DropDownButtonWidget dropdown, GameSettings s
 			dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 500, options.Keys, setupItem);
 		}
 
+		static void ShowDisplaySelectionDropdown(DropDownButtonWidget dropdown, GraphicSettings s)
+		{
+			var options = Exts.MakeArray(Game.Renderer.DisplayCount, i => "{0} {1}".F(DisplayBaseName, i + 1));
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Can we please change this method to
```csharp
static void ShowDisplaySelectionDropdown(DropDownButtonWidget dropdown, GraphicSettings s)
{
	Func<int, ScrollItemWidget, ScrollItemWidget> setupItem = (o, itemTemplate) =>
	{
		var item = ScrollItemWidget.Setup(itemTemplate,
			() => s.VideoDisplay == o,
			() => s.VideoDisplay = o);

		var label = "Display {0}".F(o + 1);
		item.Get<LabelWidget>("LABEL").GetText = () => label;
		return item;
	};

	dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 500, Enumerable.Range(0, Game.Renderer.DisplayCount), setupItem);
}
```

There are two motivations behind this request:
* There is no need to define the `options` array, and this just adds overhead resolving the array items
* The `DisplayBaseName` variable isn't needed, and hurts maintainability by introducing confusion about why it doesn't follow normal conventions more than it helps by reducing a duplicated label. All these hardcoded strings will eventually be replaced by a new translation system anyway.

**Review by pchote — COMMENTED** [review]
Looking good. Just a couple more points, then LGTM.

Can you please also squash the fixups into the first commit?

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs (hunk 11 above)
Comment (everclear72216) [reply OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Thanks I was looking for such a place.

**Comment by everclear72216** [comment]
Took care of the comments and squashed the commits. Reworded commit message.

**Comment by dragunoff** [comment]
Tested on Ubuntu, works as promised :+1: 

I'm just wondering whether the display dropdown should be visible on single monitor setups as it serves no purpose in that case :thinking:

**Comment by pchote** [comment]
We should still show it (otherwise it makes the UI code even more of a nightmare), but should disable the dropdown so its not clickable.

File: OpenRA.Game/Settings.cs [hunk 14]
```diff
@@ -176,6 +176,9 @@ public class GraphicSettings
 		[Desc("Use OpenGL ES if both ES and regular OpenGL are available.")]
 		public bool PreferGLES = false;
 
+		[Desc("Choose the monitor to use in fullscreen mode.")]
```
Comment (pchote) [inline OpenRA.Game/Settings.cs]: Style nit: Can we change this to "Display index to use in a multi-monitor fullscreen setup."? I missed this in the last round of review.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 15]
```diff
@@ -250,13 +253,19 @@ Action InitDisplayPanel(Widget panel)
...
+				ds.VideoDisplay != OriginalVideoDisplay);
 
 			var statusBarsDropDown = panel.Get<DropDownButtonWidget>("STATUS_BAR_DROPDOWN");
 			statusBarsDropDown.OnMouseDown = _ => ShowStatusBarsDropdown(statusBarsDropDown, gs);
 			statusBarsDropDown.GetText = () => gs.StatusBars == StatusBarsType.Standard ?
 				"Standard" : gs.StatusBars == StatusBarsType.DamageShow ? "Show On Damage" : "Always Show";
 
+			var displaySelectionDropDown = panel.Get<DropDownButtonWidget>("DISPLAY_SELECTION_DROPDOWN");
+			displaySelectionDropDown.OnMouseDown = _ => ShowDisplaySelectionDropdown(displaySelectionDropDown, ds);
+			var displaySelectionTextTransform = new CachedTransform<int, string>(i => "Display {0}".F(i + 1));
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Style nit: Please call this `displaySelectionLabel` to match the other `CachedTransform`s in this class.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 16]
```diff
@@ -250,13 +253,19 @@ Action InitDisplayPanel(Widget panel)
...
+			displaySelectionDropDown.GetText = () => displaySelectionTextTransform.Update(ds.VideoDisplay);
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Add `displaySelectionDropDown.IsDisabled = () => Game.Renderer.DisplayCount < 2;` to disable the dropdown if there is only a single display available.

I also just noticed that these lines are sitting between the status bar and target line dropdowns - can you please move them up to before `statusBarsDropDown`?

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/19136fb0901c6448c1d38c431fe1b761c452b4a0
